### PR TITLE
Create an ordered map for implication edges

### PR DESCRIPTION
**Pull request by yuxincs**
We have several places where we are using plain `map`s to store information. Since the map iteration order isn't guaranteed (in fact Go intentionally adds randomness to map iteration order), we are observing nondeterministic behaviors in NilAway (see more details at Issue #57).

This PR creates an ordered map as an utility and use it for storing implication edges (previously `SitesWithAssertions` and internally implemented as a `map`). More refactors will happen in follow-up PRs.

We also add tests for `inference.InferredMap` to ensure our encoding here does not increase the overall encoded binary size too much.

Depends on PR #50

### CODE REVIEW

**Review by sonalmahajan15 — COMMENTED** [review]
Overall, it makes sense to me. I have a few questions listed below.

File: inference/inferred_map.go [hunk 1]
```diff
@@ -59,16 +60,16 @@ func (*InferredMap) AFact() {}
 func (i *InferredMap) String() string {
 
 	valStr := func(val InferredVal) string {
-		switch val := val.(type) {
+		switch v := val.(type) {
```
[the remark this answers is not in the document]
Comment (yuxincs) [reply inference/inferred_map.go:63]: True! Removed 😃

File: inference/inferred_map_test.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	"go/token"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
+	"go.uber.org/nilaway/annotation"
+)
+
+// BenchmarkGobEncoding benchmarks the gob encoding of an inferred map to test the overhead.
+func BenchmarkGobEncoding(b *testing.B) {
```
[the remark this answers is not in the document]
Comment (yuxincs) [reply inference/inferred_map_test.go:15]: I guess it's not strictly related, but I used this as a preliminary performance evaluation for swapping the internal implementation of map with the ordered map. So I think it's still kinda related?

File: util/orderedmap/orderedmap.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+// we define our own `GobEncode` and `GobDecode` methods for OrderedMap).
+// Overriding the codec disallows re-uses of the same gob Encoder (stdlib does not pass the parent
+// encoder via the `GobEncode` interface), which will eventually lead to worse performance, as well
+// as larger binary size (a type will need to be encoded multiple times instead of just once if
+// the encoder is shared). See [gob encoding] for more details.
+// [gob encoding]: https://pkg.go.dev/encoding/gob
+
+// OrderedMap is an ordered map that supports iteration in insertion order. It is an _internal_
+// helper struct for NilAway and lacks some of the features of a full map.
+type OrderedMap[K comparable, V any] struct {
```
[the remark this answers is not in the document]
Comment (sonalmahajan15) [reply util/orderedmap/orderedmap.go:37]: I see. Then it's only fair to have our own implementation of an ordered map. Thanks for the explanation!

File: inference/inferred_map_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sonalmahajan15) [reply inference/inferred_map_test.go:15]: Got it, fair enough

**Review by sonalmahajan15 — APPROVED** [review]
Everything looks good

**Pull request merged**